### PR TITLE
{bintools-wrapper,cc-wrapper}: factor out Darwin SDK logic, allow paths relative to the Darwin SDK

### PR DESCRIPTION
**Pull request by emilazy**
`-L` and `-I` are interpreted relative to the `$SDKROOT` by the Darwin toolchain, so we have to avoid filtering out such paths in the purity filter hacks in order to not break e.g. the .NET Core build system. It’s also just the correct thing to do for the platform.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by emilazy** [comment]
I’ve confirmed that `dotnetCorePackages.dotnet_{8,9}.sdk` build successfully with this change without any further hacks.

File: pkgs/build-support/wrapper-common/utils.bash [hunk 1]
```diff
@@ -118,6 +118,19 @@ badPath() {
         "${p#"${TEMPDIR:-/tmp}"}" = "$p"
 }
 
+# Like `badPath`, but handles paths that may be interpreted relative to
+# `$SDKROOT` on Darwin. For example, `-L/usr/lib/swift` is interpreted
+# as `-L$SDKROOT/usr/lib/swift` when `$SDKROOT` is set and
+# `$SDKROOT/usr/lib/swift` exists.
+badPathWithDarwinSdk() {
+    sdkPath=$SDKROOT/$1
```
Comment (Artturin) [inline pkgs/build-support/wrapper-common/utils.bash:126]:
```suggestion
    sdkPath=${SDKROOT:-}/$1
```

For `set -u` (if we ever turn it on here)

Comment (emilazy) [reply pkgs/build-support/wrapper-common/utils.bash:126]: Ah, good point. Maybe I should just move the definition inside the `if`? If `@darwinMinVersion@` is set but there’s no `SDKROOT` defined then something is wrong and the logic can’t be trusted anyway, so erroring out in that case seems good to me.

Comment (emilazy) [reply pkgs/build-support/wrapper-common/utils.bash]: I’ve changed this logic to hopefully be `set -u`‐proof without silently hiding `$SDKROOT` errors in the future; let me know what you think.

**Comment by Artturin** [comment]
Building `tests.cc-wrapper.default` on linux 
Do the same on darwin

**Comment by emilazy** [comment]
I hate to be the one to tell you this, but `tests.cc-wrapper.default` doesn’t work on Darwin even on `master` 🙃 (`/nix/store/22a6p8311lg3z4ksrjbj018i7z3vgkva-clang-wrapper-16.0.6/bin/cc: line 254: mktemp: No such file or directory` – not sure why since `mktemp` is part of the `stdenv`, but I think it’s been like this for a while. I’ll take a look and see if I can fix it.)

FWIW I have built like a thousand derivations with this change to test .NET (which also goes through Swift).

**Comment by paparodeo** [comment]
> I hate to be the one to tell you this, but `tests.cc-wrapper.default` doesn’t work on Darwin even on `master` 🙃 (`/nix/store/22a6p8311lg3z4ksrjbj018i7z3vgkva-clang-wrapper-16.0.6/bin/cc: line 254: mktemp: No such file or directory` – not sure why since `mktemp` is part of the `stdenv`, but I think it’s been like this for a while. I’ll take a look and see if I can fix it.)
> 
> FWIW I have built like a thousand derivations with this change to test .NET (which also goes through Swift).

broken by https://github.com/NixOS/nixpkgs/pull/349585

the fix is in staging https://github.com/NixOS/nixpkgs/pull/349589

tho https://github.com/NixOS/nixpkgs/pull/349585 could get backed out on master and merged to staging so the cc tests work at the end of the release cycle.

**Comment by emilazy** [comment]
Oh, cool. I’ll test the merge then.

**Comment by emilazy** [comment]
FWIW we’re meant to have one more `staging` cycle before the release so it should be fine.

**Review by Artturin — APPROVED** [review]
`tests.cc-wrapper.default` passes on linux

**Review by reckenrode — APPROVED** [review]
I’ll defer to the testing done here. Regarding the Darwin flags changes, those LGTM. They’re essentially identical between cc-wrapper and bintools-wrapper, so it makes sense to consolidate them to `wrapper-common`.

**Review by philiptaron — APPROVED** [review]
This looks great and builds on Linux.

**Comment by emilazy** [comment]
It still fails because `bintools-wrapper` has copy‐pasted code from `cc-wrapper` that didn’t get the https://github.com/NixOS/nixpkgs/pull/349589 treatment :(

I will open a PR to fix that, but I just undid the tests changes locally for now, and they pass successfully 🎉

Merging on the strength of that, I hope that is okay.

**Pull request merged**